### PR TITLE
distrho-ports: 2020-07-14 -> 2021-03-15

### PR DESCRIPTION
**Pull request by tomcur**
###### Motivation for this change

Distrho-ports now contains Vitalium, a port of Vital, which is a neat new wavetable synth.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by r-rmcgibbo** [comment]
Result of `nixpkgs-review pr 116608` at 438fc4d3 run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built successfully:</summary>
  <ul>
    <li><a href="https://gist.github.com/16e86e0b83a313fcfb3ba6b55f437543">distrho</a></li></li>
  </ul>
</details>
<details>
  <summary>3 suggestions:</summary>
  <ul>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/unclear-gpl.md">unclear-gpl</a>

`gpl2` is a deprecated license, check if project uses `gpl2Plus` or `gpl2Only` and change `meta.license` accordingly.

Near pkgs/applications/audio/distrho/default.nix:81:5:
```
   |
81 |     license = with licenses; [ gpl2 gpl3 gpl2Plus lgpl3 mit ];
   |     ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/unclear-gpl.md">unclear-gpl</a>

`gpl3` is a deprecated license, check if project uses `gpl3Plus` or `gpl3Only` and change `meta.license` accordingly.

Near pkgs/applications/audio/distrho/default.nix:81:5:
```
   |
81 |     license = with licenses; [ gpl2 gpl3 gpl2Plus lgpl3 mit ];
   |     ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/unclear-gpl.md">unclear-gpl</a>

`lgpl3` is a deprecated license, check if project uses `lgpl3Plus` or `lgpl3Only` and change `meta.license` accordingly.

Near pkgs/applications/audio/distrho/default.nix:81:5:
```
   |
81 |     license = with licenses; [ gpl2 gpl3 gpl2Plus lgpl3 mit ];
   |     ^
```
</li>
  </ul>
</details>

**Comment by SuperSandro2000** [comment]
This is a semi-automatic executed nixpkgs-review with [nixpkgs-review-checks extension](https://github.com/SuperSandro2000/nixpkgs-review-checks). It is checked by a human on a best effort basis and does not build all packages (e.g. lumo, tensorflow or pytorch).
If you have any questions or problems please reach out to SuperSandro2000 on IRC.

Result of `nixpkgs-review pr 116608` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
<summary>
1 package built:
</summary>
<ul>
<li>distrho</li>
</ul>
</details>

The following issues got detected with the above build packages.
Please fix at least the ones listed with your changed packages:

<details>
distrho:
<details>


warning: unclear-gpl
`gpl2` is a deprecated license, check if project uses `gpl2Plus` or `gpl2Only` and change `meta.license` accordingly.

Near pkgs/applications/audio/distrho/default.nix:81:5:

```
   |
81 |     license = with licenses; [ gpl2 gpl3 gpl2Plus lgpl3 mit ];
   |     ^
```

See: https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/unclear-gpl.md
warning: unclear-gpl
`gpl3` is a deprecated license, check if project uses `gpl3Plus` or `gpl3Only` and change `meta.license` accordingly.

Near pkgs/applications/audio/distrho/default.nix:81:5:

```
   |
81 |     license = with licenses; [ gpl2 gpl3 gpl2Plus lgpl3 mit ];
   |     ^
```

See: https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/unclear-gpl.md
warning: unclear-gpl
`lgpl3` is a deprecated license, check if project uses `lgpl3Plus` or `lgpl3Only` and change `meta.license` accordingly.

Near pkgs/applications/audio/distrho/default.nix:81:5:

```
   |
81 |     license = with licenses; [ gpl2 gpl3 gpl2Plus lgpl3 mit ];
   |     ^
```

See: https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/unclear-gpl.md
</details>
</details>

**Comment by kanashimia** [comment]
Vitalium uses dlopen to find libfftw3f.so, otherwise it uses fallback, how would we go about this?
It can't directly link to it because of the GPL

Can we somehow wrap library? ¯×¯

To test if it uses fftw3 i did this:
```
> DEBUGGER='ltrace -e dlopen' LD_LIBRARY_PATH="/nix/store/k4la012nqx1fql8wik7rvxd57r6zjn4s-fftw-single-3.3.9/lib" ardour6
...
vitalium_lv2.so->dlopen("libfftw3f.so", 2)                      = 0x7f65f098ab60
...
```
```
> DEBUGGER='ltrace -e dlopen' ardour6
...
vitalium_lv2.so->dlopen("libfftw3f.so", 2)                      = 0
...
```

**Comment by Mic92** [comment]
> Vitalium uses dlopen to find `libfftw3f.so`, otherwise it uses fallback, how would we go about this?
> It can't directly link to it because of the GPL
> 
> Can we somehow wrap library? ¯×¯
> 
> To test if it uses fftw3 i did this:
> 
> ```
> > DEBUGGER='ltrace -e dlopen' LD_LIBRARY_PATH="/nix/store/k4la012nqx1fql8wik7rvxd57r6zjn4s-fftw-single-3.3.9/lib" ardour6
> ...
> vitalium_lv2.so->dlopen("libfftw3f.so", 2)                      = 0x7f65f098ab60
> ...
> ```
> 
> ```
> > DEBUGGER='ltrace -e dlopen' ardour6
> ...
> vitalium_lv2.so->dlopen("libfftw3f.so", 2)                      = 0
> ...
> ```

You can use patchelf on this this library to extend the RPATH. Have a look at pkgs/applications/misc/alacritty/default.nix on how to do this.

**Comment by tomcur** [comment]
I did not notice this as Vitalium appeared to work fine (aside from occasional crashes when switching between host and plugin windows, which I think is unrelated to the build process).

I've made some changes to the rpath in `postPatch`; the library now loads correctly. This seemed to be the easiest way to do it, but I'm not sure it's the correct way.

```
...
vitalium.so->dlopen("libfftw3f.so", 2)                           = 0xafe9a40
...
```

**Comment by kanashimia** [comment]
Conversation one may be interested in:
https://github.com/DISTRHO/DISTRHO-Ports/discussions/85
tldr: licence is GPLv3 only, and fftw3f is going to be linked.

My main concern before was if JUCE and FFTW were incompatible, in terms of their licences, we couldn't use patchelf i believe, as it breaks the terms.

File: pkgs/applications/audio/distrho/default.nix [hunk 1]
```diff
@@ -71,6 +89,7 @@ stdenv.mkDerivation rec {
         tal-vocoder-2
         temper
         vex
+        vitalium
         wolpertinger
     '';
     license = with licenses; [ gpl2 gpl3 gpl2Plus lgpl3 mit ];
```
Comment (SuperSandro2000) [inline pkgs/applications/audio/distrho/default.nix:95]: Can you please update them to use Only and Plus variants.

Comment (tomcur) [reply pkgs/applications/audio/distrho/default.nix]: Done. The various plugins are licensed differently (and some unclearly), so it's quite a list.

**Comment by tomcur** [comment]
I see #118928 crossed this PR. I'll create a new PR on top of #118928 to address the issues mentioned here.
